### ISSUE TITLE
WIFI Access Point: Detect incorrect password entered by station (IDFGH-13377)

### ISSUE THREAD

**Issue by koneill-txwireless** [espressif/esp-idf#14289]
### Is your feature request related to a problem?

Client has stringent cyber-security requirements and needs to be able to react, taking the matter into their own hands, to an incorrect password being entered more than once. 

### Describe the solution you'd like.

When a station attempts to connect to the ESP32 WIFI access point, with an incorrect password, a callback with a reason of "incorrect password" or something similiar. 

### Describe alternatives you've considered.

[We looked at using WIFI_EVENT_AP_STADISCONNECTED](https://www.esp32.com/viewtopic.php?f=13&t=32958) but as was correctly pointed out, that is only applicable when a station was connected in the first place. 

We tried to monitor the ESP32's reaction to an incorrect password being entered, it does log it (deauth/disassoc, reason: 3), but we can't find a way to react to it. 

![image](https://github.com/user-attachments/assets/8701b2c7-4ca3-45e8-89fd-60d8ff13ff44)

### Additional context.

The client is on v5.1.4 of IDF.

**Comment by DarmorGamz** [comment]
You've registered the ```WIFI_EVENT_AP_STADISCONNECTED``` event and logged the below in the callback?
```
wifi_event_ap_stadisconnected_t* sta_disconnected = (wifi_event_ap_stadisconnected_t*)event_data;
ESP_LOGI(DEBUG_TAG, "STA disconnect reason (%d). ", sta_disconnected->reason);
```

I'd expected a ```WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT``` value or similar, but this is how you capture the incorrect password.

**Comment by koneill-txwireless** [comment]
> You've registered the `WIFI_EVENT_AP_STADISCONNECTED` event and logged the below in the callback?
> 
> ```
> wifi_event_ap_stadisconnected_t* sta_disconnected = (wifi_event_ap_stadisconnected_t*)event_data;
> ESP_LOGI(DEBUG_TAG, "STA disconnect reason (%d). ", sta_disconnected->reason);
> ```
> 
> I'd expected a `WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT` value or similar, but this is how you capture the incorrect password.

I believe you need to be connected in the first place to get the `WIFI_EVENT_AP_STADISCONNECTED` event. Thank you though.

**Comment by KaeLL** [comment]
@koneill-txwireless no, you don't. A connection attempt with a wrong password indeed generates a WIFI_EVENT_AP_STADISCONNECTED event, which is also stated [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#wi-fi-reason-code-related-to-wrong-password).

**Comment by koneill-txwireless** [comment]
> @koneill-txwireless no, you don't. A connection attempt with a wrong password indeed generates a WIFI_EVENT_AP_STADISCONNECTED event, which is also stated [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#wi-fi-reason-code-related-to-wrong-password).

Thank you but have you actually confirmed this? Have you the log showing an incorrect password being entered? I have tested this yesterday and no event is generated when an incorrect password is entered. 

Espressif have also confirmed that `WIFI_EVENT_AP_STADISCONNECTED` cannot be used as "we can not use WIFI_EVENT_AP_STADISCONNECTED to detect failed connection attempts because WIFI_EVENT_AP_STADISCONNECTED is only posted when a successfully connected device disconnects from softAP."

**Comment by sarveshb14** [comment]
Hi @KaeLL ,
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#wi-fi-reason-code-related-to-wrong-password is application for station only.

For softAP, `WIFI_EVENT_AP_STADISCONNECTED` event is only posted when a previously successfully connected device disconnects from softAP.

**Comment by KaeLL** [comment]
I stand corrected.
The event being reported to the app layer is the one notifying about a channel change since the last connection attempt, namely `WIFI_EVENT_HOME_CHANNEL_CHANGE`, which I overlooked because right above the `esp_event` log, there's the following log from the esp-wifi-lib
`I (32711) wifi:station: 32:0e:53:a2:78:fa leave, AID = 1, reason = 15, bss_flags is 34311266, bss:0x3ffb9d04`
which means the event is being produced, but is not getting reported to the loop?

**Comment by koneill-txwireless** [comment]
Hi @sarveshb14, is this feature something you guys are considering? 

Our client is chasing us again for when we will have a solution for this. They have come from a Linux based platform to the ESP32 and it was a key feature on their legacy solution.

**Comment by sarveshb14** [comment]
Hi @koneill-txwireless , apologies for the delay. May I know which chips are you planning to use ? I will provide you with a patch.

**Comment by kriegste** [comment]
I reckon this is a helpful addition to the regular IDF on all chips.

**Comment by koneill-txwireless** [comment]
Hi @sarveshb14, apologies for being slow to get back to you, I was off the last few days. 

Your patch is working and working well. The only thing I noticed was that it appears to be triggered four times per attempt. Is there a way that I can reduce this to one per event? Should I be clearing event bits? 

![image](https://github.com/user-attachments/assets/0bd3f2af-ff02-40d7-bf77-b2c095e18937)

**Comment by sarveshb14** [comment]
Hi @koneill-txwireless , for WPA2, we (softAP) know that station is trying to connect with wrong password if mic-verification (kind of signature verification) fails during the processing of M2 message (of 4-Way-Handshake). Its normal for stations to re-transmit frames in case of failed connection in the assumption that the frame was previously lost in the transmission. 

Hence you are seeing the event getting triggered for 4 times. It is actually the station which is re-transmitting the frame for 4 times (due to wrong password). This can also happen for WPA3.

I do not think we can do much here. We have to report the event every time a wrong frame (due to wrong password) is received.

**Comment by koneill-txwireless** [comment]
That's fine, thank you for the background on it. We can manage the multiple triggers.

**Comment by KaeLL** [comment]
@sarveshb14 What's the ETA for that on master?

**Comment by koneill-txwireless** [comment]
Hi @sarveshb14, is there a possibility that this will be merged into the next release of v5.1, e.g. v5.1.5? It is the release line that our client is on and they are sensitive to big jumps.

**Comment by koneill-txwireless** [comment]
Hey @sarveshb14, any update on this? 

Client is keen to get this integrated and tested.

**Comment by koneill-txwireless** [comment]
Since this was requested, v5.1.5 has been released. The libraries provided are targetted at v5.1.4 I think as you get linker errors on v5.1.5. Would it be possible to get new WIFI libs with incorrect password detection functionality for v5.1.5?

**Comment by sarveshb14** [comment]
Hi @koneill-txwireless , Sorry for the late response. 

The functionality to detect the station using wrong pasword/pin is implemented completely in `wpa_supplicant` component of `esp-idf`. You can omit the library change and cherry-pick the commit directly on v5.1.5.  You will still be able to compile your application without any linking errors. (library change is just to make sure that the version of public header file `components/esp_wifi/include/esp_wifi_types.h` is not different in esp-idf and libraries)

We will get this feature merged in esp-idf soon. Till then please use this patch as it is on esp-idf (without the libraries).

**Comment by Alvin1Zhang** [comment]
Thanks for reporting, the fix is available at https://github.com/espressif/esp-idf/commit/fae82cfaa0b0f283b68c339057889f34129cb915.

**Comment by Alvin1Zhang** [comment]
Thanks for reporting, and sorry for the slow turnaround, fix has been back ported to branch until release/5.1, and the fix on release/5.1 is available at https://github.com/espressif/esp-idf/commit/39fe10677ca7e43dc9a7075c290a177f09345cf4, and is included in the v5.1.6 release (https://github.com/espressif/esp-idf/releases/tag/v5.1.6), feel free to reopen if the issue still happens.

**Comment by sarveshb14** [comment]
Hi @koneill-txwireless , it is merged in release/v5.4 and will be available on the upcoming tag v5.4 (v5.4.1)

**Comment by koneill-txwireless** [comment]
Excellent, thank you.

**Comment by koneill-txwireless** [comment]
Hi, @sarveshb14, is there an estimated date for release of the upcoming tag? We have a client that requires it.

**Comment by Alvin1Zhang** [comment]
@koneill-txwireless Thanks for reporting, regarding v5.4.1, per https://github.com/espressif/esp-idf/blob/master/ROADMAP.md#esp-idf-bugfix-releases, v5.4.1 will release soon, will update here once v5.4.1 releases, thanks.
